pkg/offsetguess/offsetguess.go

```diff
@@ -0,0 +1,304 @@
+			if err != nil {
+				return fmt.Errorf("error converting source port: %v", err)
+			}
+
+			sport = int(htons(uint16(sport)))
+
+			// set SO_LINGER to 0 so the connection state after closing is
+			// CLOSE instead of TIME_WAIT. In this way, they will disappear
+			// from the conntrack table after around 10 seconds instead of 2
+			// minutes
+			if tcpConn, ok := conn.(*net.TCPConn); ok {
+				tcpConn.SetLinger(0)
+			} else {
+				return fmt.Errorf("not a tcp connection unexpectedly")
+			}
+
+			conn.Close()
+		} else {
+			conn, err := net.Dial("tcp6", fmt.Sprintf("[%s]:9092", ip))
+			if err == nil {
+				conn.Close()
+			}
+		}
+
+		err = b.LookupElement(mp, unsafe.Pointer(&zero), unsafe.Pointer(&status))
+		if err != nil {
+			return fmt.Errorf("error reading tcptracer_status: %v", err)
+		}
+
+		if status.status == checked {
+			switch status.what {
+			case guessSaddr:
+				if status.saddr == uint32(saddr) {
+					status.what++
+					status.status = checking
+				} else {
+					status.offsetSaddr++
+					status.status = checking
+					status.saddr = uint32(saddr)
+				}
+			case guessDaddr:
+				if status.daddr == uint32(daddr) {
+					status.what++
+					status.status = checking
+				} else {
+					status.offsetDaddr++
+					status.status = checking
+					status.daddr = uint32(daddr)
+				}
+			case guessFamily:
+				if status.family == uint16(family) {
+					status.what++
+					status.status = checking
+					// we know the sport ((struct inet_sock)->inet_sport) is
+					// after the family field, so we start from there
+					status.offsetSport = status.offsetFamily
+				} else {
+					status.offsetFamily++
+					status.status = checking
+				}
+			case guessSport:
+				if status.sport == uint16(sport) {
+					status.what++
+					status.status = checking
+				} else {
+					status.offsetSport++
+					status.status = checking
+				}
+			case guessDport:
+				if status.dport == dport {
+					status.what++
+					status.status = checking
+				} else {
+					status.offsetDport++
+					status.status = checking
+				}
+			case guessNetns:
+				if status.netns == netns {
+					status.what++
+					status.status = checking
+				} else {
+					status.offsetIno++
+					// go to the next offsetNetns if we get an error
+					if status.err != 0 || status.offsetIno >= 200 {
+						status.offsetIno = 0
+						status.offsetNetns++
+					}
+					status.status = checking
+				}
+			case guessDaddrIPv6:
+				if compareIPv6(status.daddrIPv6, daddrIPv6) {
+					status.what++
+					status.status = ready
+				} else {
+					status.offsetDaddrIPv6++
+					status.status = checking
+				}
+			default:
+				return fmt.Errorf("unexpected field")
+			}
+		}
+
+		if err := b.UpdateElement(mp, unsafe.Pointer(&zero), unsafe.Pointer(&status), 0); err != nil {
+			return fmt.Errorf("error updating tcptracer_status: %v", err)
+		}
+
+		if status.offsetSaddr >= 200 || status.offsetDaddr >= 200 ||
```

Comment:
In fact, couldn't an out-of-bounds access on the socket DS potentially cause a crash on the ebpf probe?